TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot/subsystems/XDrive.java

```diff
@@ -0,0 +1,267 @@
+import org.firstinspires.ftc.teamcode.robot.constants.DrivetrainBuilderConstants;
+import org.firstinspires.ftc.teamcode.robot.constants.XDriveConstants;
+import org.opencv.core.Mat;
+
+public class XDrive implements Subsystem {
+    enum DrivetrainState{
+        MANUAL_CONTROL,
+        APRIL_TAG_ALIGNMENT
+    }
+    public DrivetrainState currentDriveState = DrivetrainState.MANUAL_CONTROL;
+    private  Telemetry telemetry;
+    private static XDrive instance;
+    private double reset_angle = 0;
+    private DcMotor front_left = null;
+    private DcMotor back_left = null;
+    private DcMotor back_right = null;
+    private DcMotor front_right = null;
+    private IMU imu;
+
+
+    private PIDController XpositionPID=AutonomousConstants.stopAtPointPID;
+    private PIDController YpositionPID=AutonomousConstants.stopAtPointPID;
+    private PIDController HpositionPID=AutonomousConstants.stopAtPointPID;
+
+    Pose2d currentPose = START_POSITION;
+    int prevFLTicks1 = 0;
+    int prevFRTicks1 = 0;
+    int prevBLTicks1 = 0;
+    int prevBRTicks1 = 0;
+    int prevFLTicks2 = 0;
+    int prevFRTicks2 = 0;
+    int prevBLTicks2 = 0;
+    int prevBRTicks2 = 0;
+
+    private XDrive() {
+    }
+
+    @Override
+    public void initialize(HardwareMap hardwareMap, Telemetry telemetry) {
+        front_left = hardwareMap.dcMotor.get(MOTOR_FRONT_LEFT);
+        back_left = hardwareMap.dcMotor.get(MOTOR_BACK_LEFT);
+        back_right = hardwareMap.dcMotor.get(MOTOR_BACK_RIGHT);
+        front_right = hardwareMap.dcMotor.get(MOTOR_FRONT_RIGHT);
+
+        front_left.setDirection(DcMotor.Direction.REVERSE);
+        back_left.setDirection(DcMotor.Direction.REVERSE);
+        front_right.setDirection(DcMotor.Direction.FORWARD);
+        back_right.setDirection(DcMotor.Direction.FORWARD);
+
+        front_left.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+        back_left.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+        front_right.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+        back_right.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
+
+        this.telemetry = telemetry;
+
+        imu = hardwareMap.get(IMU.class, "imu");
+
+        telemetry.addLine("Xdrive initialized");
+    }
+
+    @Override
+    public void start() {
+    }
+
+    @Override
+    public void stop() {
+    }
+
+    @Override
+    public void execute(GamepadManager gamepadManager) {
+        controlDriveState(gamepadManager.getDriver());
+        if (currentDriveState == DrivetrainState.MANUAL_CONTROL) {
+
+            drive(gamepadManager.getDriver());
+            resetAngle(gamepadManager.getDriver());
+        }
+    }
+    public void controlDriveState(SmartGamepad driver){
+        if (Math.abs(driver.getLeftStickY())> XDriveConstants.APRIL_TAG_BREAK_TOLERANCE ||Math.abs(driver.getRightStickX())>XDriveConstants.APRIL_TAG_BREAK_TOLERANCE){
+            currentDriveState = DrivetrainState.MANUAL_CONTROL;
```

Comment:
using POO it's not a good approach to change a variable state in a void method. 
I recommend creating the variable here and then returning it's value

Reply:
yeah that makes a lot more sense than what i was doing